### ISSUE TITLE
Exposing index inspection/manipulation macros/functions

### ISSUE THREAD

**Issue by zachasme** [uber/h3#277]
I'm implementing (SP-)GiST index support for our PostgreSQL bindings and I find myself copying over macros for inspecting and manipulating index digits and resolution. I need these because I'm encoding each index digit as a branch in a tree, and reconstructing H3 indexes as the path from root to leaf node. 

Have you considered exposing some of the tools for low-level manipulation of H3 indexes as part of the public API or maybe in a separate header file?

**Comment by dfellis** [comment]
Is there a problem with the `h3-pg` project depending on the `h3Index.h` file? I think everything you need is defined in that header file.

**Comment by zachasme** [comment]
Yes those are exactly the things I need. The only problem was that I couldn't figure out a way to install `h3Index.h` alongside `h3api.h`, and I would like to support linking against H3 built as a shared library at some point.

**Comment by isaacbrodsky** [comment]
To install `h3Index.h`, you can modify `CMakeLists.txt` before generating and `make install`ing:

Change (around line 609):
```
install(
    FILES "${CMAKE_CURRENT_BINARY_DIR}/src/h3lib/include/h3api.h"
    DESTINATION "${include_install_dir}/h3"
)
```

To:
```
install(
    FILES "${CMAKE_CURRENT_BINARY_DIR}/src/h3lib/include/h3api.h" "${CMAKE_CURRENT_SOURCE_DIR}/src/h3lib/include/h3Index.h"
    DESTINATION "${include_install_dir}/h3"
)
```

I don't think that linking to H3 as a shared library should have an impact on using the macros or using internal H3 functions.

**Comment by dfellis** [comment]
Should that be a default, then @isaacbrodsky ?

**Comment by zachasme** [comment]
Unless it is on by default I cannot assume existence of `h3Index.h` even if `h3` is already installed on the system.

For example on Arch Linux `h3` might have been installed manually or using [the aur package](https://aur.archlinux.org/packages/h3-git/) but I will have to instruct users to rebuild it using the changed `CMakeLists`.

I fully understand if it doesn't make sense for you to expose that file by default in it's current form.

**Comment by isaacbrodsky** [comment]
Only `h3api.h` was installed because that's the public API. I think it would be fine to add the internal headers with clarity that they're internal. The drawback of this is that different library versions may break your build. Perhaps we could initially add something like:

```
# (First, define internal_header_files)
install(
    FILES ${internal_header_files}
    DESTINATION "${include_install_dir}/h3/internal"
)
```

 Alternately maybe the macros could be promoted to part of the public API, and add them to `h3api.h` or  a new, non-internal, header that's installed. The bitwise layout of the indexes is already part of the public API so that would be a small change.

**Comment by dfellis** [comment]
I think moving the macros to `h3api.h` is the best solution here.

**Comment by zachasme** [comment]
Would you accept a PR moving some of the macros into `h3api.h`?

**Comment by isaacbrodsky** [comment]
> Would you accept a PR moving some of the macros into `h3api.h`?

Yes, I think that would be accepted.

**Comment by dfellis** [comment]
No blocking from me, sorry that I forgot about this.

**Comment by ajfriend** [comment]
@zachasme, I'm curious. What are you using the tree representations for?

**Comment by zachasme** [comment]
@ajfriend I built an H3-specific indexing scheme (sp-gist) for PostgreSQL. When encoding the index I can save space by only storing/retrieving one index digit per node/branch.

For basic lookups nothing beats a good old b-tree, but if you are querying by something like containment or overlap, a specialized search tree will be much faster.

**Comment by ajfriend** [comment]
@zachasme, cool! Is it [this](https://pgxn.org/dist/h3/docs/api.html#SP-GiST.operator.class.experimental.)? Do you have any examples or benchmarks? I'd be super interested to see what kinds of speedups you see!

**Comment by zachasme** [comment]
@ajfriend yes, that's it! To be frank, the project it was meant for was discontinued before I had a chance to properly benchmark it though.

In *theory* it should be faster on large datasets for queries such as `SELECT avg(amount) FROM cells WHERE cell <@ '811f3ffffffffff';`
